Incremental unwinding of one specified loop -- rebase

xbauch opened the pull request:
Incremental unwinding and assertion checking of a specified loop.

This is a rebase of #4361.

- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

peterschrammel committed:
Add single-loop incremental symex checker

Checks the assertions after each iteration of
a specified loop.

peterschrammel committed:
Add options for incremental one-loop BMC

for incremental unwinding and assertion checking
of a specified loop.

peterschrammel committed:
Allow symex to temporarily ignore assertions

Allows ignoring assertions until a certain condition
is satisfied (e.g. a certain unwinding is reached).
Used for implementation of ignore-assertions-before-unwind-min.

peterschrammel committed:
Add --incremental-loop to CBMC

for incremental unwinding and assertion checking
of a specified loop.

peterschrammel committed:
Allow symex_target_equationt to separately convert assertions

This enables the caller to handle assertion conversion differently,
e.g. for incremental loop unwinding.

peterschrammel committed:
Add callback to decide on interrupting goto-symex

Required for incremental loop unwinding.

peterschrammel committed:
Add symex BMC incremental one loop

Implement callback that decides on when to interrupt
goto-symex during incremental unwinding.

peterschrammel committed:
Clean up cbmc-incr-oneloop tests

Option name has changed.

peterschrammel committed:
Fix quoting in test command

Otherwise single quotes get incorrectly stripped off.

peterschrammel committed:
Use only remaining properties in property decider

We can ignore already failed properties.

peterschrammel committed:
Allow disabling optimized assertion conversion

Cannot be used with incremental loop unwinding.

peterschrammel committed:
Hidden field is used for hiding in trace

not 'ignore' (which is used for slicing).

peterschrammel committed:
Ignore properties before unwind min

Allows to check assertions only between unwind-min and unwind-max
iterations of a specified loop.

peterschrammel committed:
Set induction tests to FUTURE

These require more options that have not been implemented yet.

peterschrammel committed:
Clang-format cbmc-incr-oneloop tests

peterschrammel committed:
Make sure output buffers are flushed before writing exit code

This seems to work when calling bash again. Otherwise,
we can get something like:

This is half a liEXIT=123

hannes-steffenhagen-diffblue commented:
@tautschnig Were your comments from https://github.com/diffblue/cbmc/pull/4361 addressed?

thk123 commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -318,6 +318,19 @@ void symex_target_equationt::constraint(
   merge_ireps(SSA_step);
 }
 
+void symex_target_equationt::convert_without_assertions(
```
> Could this be refactored so that we don't have almost-the-same-code twice?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292643973

Looks to be outstanding

xbauch replied:
Refactored.

thk123 commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -458,13 +472,13 @@ void symex_target_equationt::convert_assertions(
       // ignore already converted assertions in the error trace
       if(step.is_assert() && step.converted)
-        step.ignore = true;
+        step.hidden = true;
```
> That makes the comment above seem rather odd?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292549222

Outstanding

xbauch replied:
Comment changed (also below).

thk123 commented on src/goto-symex/symex_target_equation.cpp:
```diff
@@ -491,7 +505,7 @@ void symex_target_equationt::convert_assertions(
     // ignore already converted assertions in the error trace
     if(step.is_assert() && step.converted)
-      step.ignore = true;
+      step.hidden = true;
```
Ditto

thk123 commented on src/goto-symex/goto_symex.h:
```diff
@@ -164,6 +164,17 @@ class goto_symext
+  /// by the symbolic executions.
+  bool ignore_assertions = false;
+
+  virtual bool check_break(
```
> Documentation?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292644774

outstanding

xbauch replied:
Documented (and refactored a bit).

thk123 commented on regression/cbmc-incr-oneloop/Makefile:
```diff
@@ -1,10 +1,10 @@
 
 tests.log: ../test.pl
-	@../test.pl -c "perl -e 'alarm shift @ARGV; exec @ARGV' 30 ../../../src/cbmc/cbmc --slice-formula"
+	@../test.pl -c "perl -e 'alarm shift @ARGV; exec @ARGV' 8 ../../../src/cbmc/cbmc --slice-formula"
```
Michael would like a comment here explaining the perl invocation is for the timeout

xbauch replied:
Comment added.

thk123 commented on src/goto-checker/symex_bmc_incremental_one_loop.cpp:
```diff
@@ -0,0 +1,135 @@
+                                   : 0)
+{
+  ignore_assertions =
+    options.get_bool_option("ignore-properties-before-unwind-min");
```
Perhaps a question for @peterschrammel 

> When was this option introduced and where is it documented?

https://github.com/diffblue/cbmc/pull/4361#discussion_r292647453

peterschrammel replied:
It has been introduced as part of this PR. Maybe it hasn't been documented as part of thIs PR, though.

thk123 replied:
Hmm, seems to be documented here @tautschnig : https://github.com/diffblue/cbmc/pull/5217/files#diff-e58a25459f0e8ca5a6c351b1d9d75167R225

xbauch replied:
```suggestion
    incr_min_unwind >= 2 && options.get_bool_option("ignore-properties-before-unwind-min");
```
@peterschrammel or at least this way it was in 5.4.

peterschrammel replied:
Any observable effect of adding this?

xbauch replied:
Well since the test(s) in question do not set the `unwind-min` I assume it is 1 by default. So the ignoring would not even occur. So yes, the test passes, but rather trivially.

xbauch replied:
On the other hand, the test is still wrong if you set the `unwind-min` (to 10 for example) even with the suggested change.

thk123 reviewed (commented):
Comments from @tautschnig on original PR that are still outstanding

thk123 commented:
Niavely looking at coverage we should add the following tests:
 - a pretty shortest trace (e.g. with `--beautify` turned on for one of the tests)
 - building a non-shortest trace for a property
 - building a full trace

(dunno how you exercise those last two options) 

Also, we should add a trivial test for  checking exit with non-zero exit code if ran with --paths as well as --incremental-loop just to stop any boneheaded mistakes in future 

Also, less nit picky, but it seems like `symex_bmc_incremental_one_loopt` is not tested with any `loop_unwind_handlers`. This seems like it should be added

thk123 commented on src/goto-symex/goto_symex.h:
```diff
@@ -164,6 +164,22 @@ class goto_symext
+  /// \param unwind: current unwinding counter
+  /// \return true if the symbolic execution is to be interrupted for checking
+  virtual bool check_break(
+    const irep_idt &id,
```
:pick: is this the loop id? given how frequent `id`s are used in this code base, I'd probably call it `loop_id`

xbauch replied:
Done.

thk123 reviewed (approved):
AFAICT all of @tautschnig comments from #4361 have been addressed.

xbauch commented:
> * a pretty shortest trace (e.g. with `--beautify` turned on for one of the tests)

It seems that the refinement solver does not support beatification

xbauch commented:
> * a pretty shortest trace (e.g. with `--beautify` turned on for one of the tests)
> * building a non-shortest trace for a property
> * building a full trace

I have a suspicion that some of the `build_*trace` functions are only there to override the default and cannot really be called in any of our factory-produced verifiers. I'll keep digging.

petr-bauch committed:
Add documentation and refactor incremental one loop changes

`check_break` is now documented in `goto_symex.h` as well (I removed the
parameters that were not used anywhere).

xbauch commented:
@peterschrammel In an attempt to improve coverage, the last commit adds a test for trace production. I've also suggested commenting out the implementation of the "beautified" trace and the "full" trace since I don't think they can be exercised by any of the current verifier combinations. Is it change acceptable?

peterschrammel commented on src/goto-checker/single_loop_incremental_symex_checker.cpp:
```diff
@@ -168,25 +168,33 @@ operator()(propertiest &properties)
-    goto_trace);
-
-  return goto_trace;
+  UNREACHABLE;
```
🚫 Write a test for it if you want to cover it.

xbauch replied:
Done.

petr-bauch committed:
Add tests for traces in one-loop checker

1: working test for the standard trace.
2: failing test for the beautified shortest trace.

thk123 commented on regression/cbmc-incr-oneloop/ignore-before-unwind/main.c:
```diff
@@ -0,0 +1,14 @@
+int main()
```
:pick: typo in commit message `KNOWBUG` -> `KNOWNBUG`

thk123 commented on regression/cbmc-incr-oneloop/alarm2/fancy_trace_test.desc:
```diff
@@ -0,0 +1,12 @@
+--
+^warning: ignoring
+--
+This will fail the invariant on `build_goto_trace.cpp:320` asserting certain
```
Could you raise a bug on cbmc and put the issue number here.

thk123 commented on regression/cbmc-incr-oneloop/alarm2/trace_test.desc:
```diff
@@ -0,0 +1,9 @@
+^VERIFICATION FAILED$
+^State \d+ file main.c function havocIO line 501 thread \d+$
+--
+^warning: ignoring
```
A description of why you added this test (e.g. tests the trace production for ??)

thk123 commented:
Maybe you already knew this, but the key flag is ` --ignore-properties-before-unwind-min`, disabling that causes it to find the failure

peterschrammel commented on regression/cbmc-incr-oneloop/ignore-before-unwind/main.c:
```diff
@@ -0,0 +1,14 @@
+
+  while(1)
+  {
+    __CPROVER_assume(property == 1);
```
What is the purpose of the unused `assumption` variable?

xbauch replied:
Hard to say, but if you comment it out, it suddenly starts to work.

peterschrammel replied:
I mean the assumption variable, not the assumption statement.

xbauch replied:
Sorry, yes, that one is irrelevant.

petr-bauch committed:
Fix ignoring properties before unwind min (off-by-one)

Basically, if unwind-min is zero (by default) the first loop iterations should
not be ignored (even if requested). Tests show the problem.

xbauch commented:
@peterschrammel I've added the check for when to start ignoring properties (eebfcd98675f6392b6ec962481f5465c543ddb40) and cleaned up the test.

martin-cs reviewed (approved):
If this branch has the history I think it does it might be a candidate for the longest running branch as it predates even VSD!

thk123 commented:
Just noting the travis job has passed, but for some reason the icon has not refreshed on here

peterschrammel commented on regression/cbmc-incr-oneloop/unwindset-more-loops1/test.desc:
```diff
@@ -0,0 +1,13 @@
+^Unwinding loop main\.3 iteration 3
+--
+^warning: ignoring
+^Unwinding loop main\.0 iteration 6
```
Missing new line at end of file

xbauch replied:
Fixed (I had to rerun the CI anyway).

petr-bauch committed:
Add missing newline symbol

to a test description file.
